### DISCUSSION TITLE
Test Matrix of Hammer DB 3.2 or Hammer DB 3.1

### DISCUSSION THREAD

**Discussion by b-anup** [TPC-Council/HammerDB#117] [Q&A] [answered]
I am looking for Database test matrix specifically for SQL 2019 .
I can see it for V 3.3 but not able to get it for v 3.1 & v 3.2 .

So i was wondering if SQL 2019 has been tested for any of the Hammer DB version.

**Comment by sm-shaw** [comment]
The test matrix is updated for each release so will be updated for the next version. Having said that SQL 2019 on Linux has been tested with v3.3 and no specific issues have been identified. It is strongly recommended to use v3.3, in particular this changed the underlying interface from ODBC v1 to ODBC v3 and resolved Issue #72 - also there was a fix for a NUMA issue where prior to and including v3.2 connections could be made to one NUMA node.  The results should not be directly compared between versions - in testing v3.3 performs slightly better than v3.2 as a result of the updated ODBC interface. If you do compare v3.3 results with earlier versions then only NOPM should be used as the TPM values are different due to a change in the number of commits as a result of changing the interface.

**Comment by b-anup** [comment]
Well i was referring this https://www.hammerdb.com/docs/ch01s02.html and got confuse.
Thanks for clarifying it .

There is a huge difference for V 3.1 and V3.3 transaction numbers.
V 3.1 gives more TPM than V 3.1 .So wanted to clarify which one to refer.

**Comment by sm-shaw** [comment, marked as answer]
Yes, that matrix is the versions that were specifically tested before release, however that doesn't mean that other versions are not supported or even have not been tested since release. Yes for SQL Server between v3.2 and v3.3 the compiled interface between HammerDB and ODBC changed from tclodbc to tdbc::odbc (Note tdbc::odbc is modified from what is available included with the standard TCL release).  With tclodbc (up to 3,.2) there was an additional external commit to the stored procedures - you can see this line as "`odbc commit`".  However if this was removed or commented out the sessions would hang with the error “`connection is busy with results for another command`”.   The newer interface did not need this additional commit so it was removed. This means that comparing the TPM means that in v3.1/3.2 the reported TPM (ie batches/sec) is much higher however the actual work done (NOPM) is very similar (with v3.3 marginally faster).  The NUMA bug that was fixed is described here https://www.n0derunner.com/2019/10/sql-server-uses-only-one-numa-node-with-hammerdb/.  It is always best practice to compare testing with one version with the same version - as this is an example where changes to the reported metrics were unavoidable as a result of code changes.

**Comment by sm-shaw** [comment]
closed as no further update
